File: julia/Day13.jl
module Day13

export Submarine, on_input!, output

@enum ReadState ReadingDots ReadingFoldingInstructions

mutable struct Submarine
    input_read_state::ReadState
    dots::Vector{Tuple{UInt32, UInt32}}
    num_points_after_first_fold::UInt32

    Submarine() = new(ReadingDots, [], 0)
end

function on_input!(s::Submarine, a::AbstractString)
    on_input!(s::Submarine, Val(s.input_read_state), a)
end

function on_input!(s::Submarine, ::Val{ReadingDots}, a::AbstractString)
    if a == ""
        s.input_read_state = ReadingFoldingInstructions
        return
    end

    x, y = collect(parse.(UInt32, split(a, ",")))
    push!(s.dots, (x, y))
end

function on_input!(s::Submarine, ::Val{ReadingFoldingInstructions}, a::AbstractString)
    axis, coord = collect(split(chop(a, head = length("fold along "), tail = 0), "="))
    coord = parse(UInt32, coord)

    fold!(s, axis, coord)
end

function fold!(s::Submarine, axis, coord)
    if axis == "x"
        folded_points = collect(map(p -> (coord - (p[1] - coord), p[2]), filter(p -> p[1] > coord, s.dots)))
        filter!(p -> p[1] < coord, s.dots)
        s.dots = [s.dots; folded_points]
    else
        folded_points = collect(map(p -> (p[1], coord - (p[2] - coord)), filter(p -> p[2] > coord, s.dots)))
        filter!(p -> p[2] < coord, s.dots)
        s.dots = [s.dots; folded_points]
    end

    unique!(s.dots)

    if s.num_points_after_first_fold == 0
        s.num_points_after_first_fold = length(s.dots)
    end
end

function transpose_tuple((x, y))
    (y, x)
end

function print_dots(s::Submarine)
    max_x = maximum(map(p -> p[1], s.dots))
    max_y = maximum(map(p -> p[2], s.dots))

    sorted_dots = sort(s.dots, by = transpose_tuple)
    # next is a tuple `(val, state)` or `nothing`.
    next = iterate(sorted_dots)

    for y = 0:max_y
        for x = 0:max_x
            if next !== nothing && (x, y) == next[1]
                print("#")
                next = iterate(sorted_dots, next[2])
            else
                print(" ")
            end
        end

        println()
    end
end

function output(s::Submarine)
    println("Part 1: $(s.num_points_after_first_fold)")
    println("Part 2:")
    print_dots(s)
end

end

File: rust/src/day13.rs
enum InputReadState {
    ReadingDots,
    ReadingFoldInstructions,
}

pub struct Submarine {
    input_read_state: InputReadState,
    dots: Vec<(u32, u32)>,
    num_folds: u32,
}

impl Submarine {
    pub fn new() -> Self {
        Self {
            input_read_state: InputReadState::ReadingDots,
            dots: Vec::new(),
            num_folds: 0,
        }
    }

    pub fn on_input(&mut self, result: &str) {
        match self.input_read_state {
            InputReadState::ReadingDots => {
                if result == "" {
                    self.input_read_state = InputReadState::ReadingFoldInstructions;
                    return;
                }

                let mut coords = result.split(",").map(|x| x.parse().unwrap());
                let x: u32 = coords.next().unwrap();
                let y: u32 = coords.next().unwrap();

                self.dots.push((x, y));
            }
            InputReadState::ReadingFoldInstructions => {
                // Needed for part 1.
                if self.num_folds == 1 {
                    return;
                }

                let mut instructions = result.strip_prefix("fold along ").unwrap().split("=");
                let axis = instructions.next().unwrap();
                let coord: u32 = instructions.next().unwrap().parse().unwrap();

                self.fold(axis, coord);
            }
        }
    }

    fn fold(&mut self, axis: &str, coord: u32) {
        self.num_folds += 1;

        match axis {
            "x" => {
                let points_beyond_fold_line = self.dots.iter().filter(|(x, _)| *x > coord);
                let mut folded_points: Vec<(u32, u32)> = points_beyond_fold_line.map(|(x, y)| (coord - (*x - coord), *y)).collect();
                self.dots.retain(|(x, _)| *x < coord); // Always guaranteed that the fold line won't have dots, so we can ignore cases where `*x == cord`.
                self.dots.append(&mut folded_points);
            }
            "y" => {
                let points_beyond_fold_line = self.dots.iter().filter(|(_, y)| *y > coord);
                let mut folded_points: Vec<(u32, u32)> = points_beyond_fold_line.map(|(x, y)| (*x, coord - (*y - coord))).collect();
                self.dots.retain(|(_, y)| *y < coord); // Always guaranteed that the fold line won't have dots, so we can ignore cases where `*y == cord`.
                self.dots.append(&mut folded_points);
            }
            _ => panic!("Got an invalid axis!")
        }

        self.dots.sort_unstable();
        self.dots.dedup();
    }

    pub fn output(&mut self) {
        println!("Part 1: {}", self.dots.len());
    }
}